### ISSUE TITLE
AP + captive portal?

### ISSUE THREAD

**Issue by michal2229** [nodemcu/nodemcu-firmware#59]
Hi there,

I do not know is it a good place to ask this, but i wonder if there is any simple way of implementing captive portal on ESP? I mean something like the login page opened automatically after connecting to wifi hotspot in a hotel room. 
I am not very experienced in network stuff like this... 

Thanks in advance.

**Comment by layerzerolabs** [comment]
I've been attempting this, and have sucessfully compiled nodeMCU with 'USE_DNS' defined in dhcp.h. This then gives the esp ip as a dns server in the dhcp options.

However, there is a bug reported by ThomasW here: http://www.esp8266.com/posting.php?mode=reply&f=24&t=1285#pr7590
"Interestingly, i just learned that if you stop a server as described above and start another one on a different port the module keeps listening on the previous port too (I think this is a bug) but uses the callbacks of the active one."
If this bug can be reproduced and then fixed, we might be able to create a dns server, respond to a request, close down the server and then start a http server process.
Even better would be the ability to run two server processes (on different ports) at the same time?

**Comment by acalatrava** [comment]
It seems you can start a TCP server AND a UDP server. You can set the DNS server in UDP since the client will use UDP (unless the request or the response is too large)

**Comment by zendes** [comment]
You can't hijack DNS and just answer with the ESP's IP address in every query, because the client will cache them. After they've gone through the portal, how will you know where to forward the request?

Example:
client requests www.facebook.com
you answer with the ip of the esp
client connects to you
they have a valid password so you let them through
*problems start here*
client still thinks www.facebook.com is at the ip of the esp, so it reconnects to you
you don't know what hostname the user typed in, so you don't know where to forward them to...

**Comment by acalatrava** [comment]
Well, that could be easily done by setting a very low TTL on every reply. 

However I don't think that's the problem since I guess that the captive portal will be only for configuration. Once the configuration is done the captive portal will be turned off.

**Comment by zendes** [comment]
why would you use a captive portal for that? those are for intercepting traffic that's not destined for you. if you want to just redirect every single dns name to your esp so you land on a config page, why not type in it's ip address? I don't think that's very different from either giving someone a custom dns name to use, or explaining that they can surf to www.facebook.com to configure their esp.

If you want to push this even further, we should implement avahi on top of it so you can address all of them by name even if they're wifi clients. :-)

**Comment by acalatrava** [comment]
Yes I know, but why would you use a captive portal for anything else on this chip? Ok, you capture the user, ask for a user & password, and then what? Use a reverse proxy? I don't think this is as powerful as that... So the only scenario I can think of is to use the captive portal for a configuration portal...

But you are right, there is no difference to say "point to 192.168.4.1 to configure the xxx" or "point to any web page to configure the xxx"...

**Comment by layerzerolabs** [comment]
As zendes and acalatrava have commented, this is only good for intercepting traffic - and equivalent to telling users to type in an IP address. For technologically savvy users this isn't a problem, but many 'ordinary' users have no idea what an IP address is, that the punctuation is dots not commas etc.
Also browsers sometimes do funny things especially with omniboxes and substitute an internet search for www.192.168.4.1.com or somesuch.
It just means one less technical detail to explain, and more or less guarantees that the user will be presented with the configuration page. As soon as the ESP has joined the main network as a client, the AP is switched off so that with any luck the client will auto-reconnect to the main network, get a new gateway and a fresh set of dns servers, so the issues of forwarding traffic or caching the names shouldn't be a problem.

**Comment by mtnbrit** [comment]
Any further progress on this? Would be great if there was a way to tell the softAP DHCP server to issue its own IP as DNS server without compiling nodeMCU?

**Comment by mtnbrit** [comment]
Heres a DNS server that always responds with 192.168.4.1

```
function unhex(str)
    str = string.gsub (str, "(%x%x) ?",
        function(h) return string.char(tonumber(h,16)) end)
    return str
end

s=net.createServer(net.UDP)
s:on("receive",function(s,c) 

transaction_id=string.sub(c,1,2)
flags=string.sub(c,3,4)
questions=string.sub(c,5,6)

query = ""
raw_query = ""
j=13
while true do
    byte = string.sub(c,j,j)
    j=j+1
    raw_query = raw_query .. byte
    if byte:byte(1)==0x00 then --NULL marks end of the string.
        break
    end
    for i=1,byte:byte(1) do
        byte = string.sub(c,j,j)
        j=j+1
        raw_query = raw_query .. byte
        query = query .. byte
    end
    query = query .. '.'
end
query=query:sub(1,query:len()-1) --strip the trailing dot.
q_type = string.sub(c,j,j+1)
j=j+2
if q_type == unhex("00 01") then 
--print("Got a type A query "..query)
class = string.sub(c,j,j+1)

--ip = "192.168.4.1"
ip=unhex("C0 A8 04 01")
answers = unhex("00 01")
flags = unhex("81 80")

resp=transaction_id..flags..questions..answers..unhex("00 00")..unhex("00 00")..raw_query..q_type..class
resp=resp..unhex("c0 0c")..q_type..class..unhex("00 00 00 da")..unhex("00 04")..ip
s:send(resp)
end
end) 
s:on("sent",function(s) 
--s:close() 
end)
s:listen(53)
print("listening, free:", node.heap())
```

Credit to Jacek Wielemborek for this posting: http://seclists.org/nmap-dev/2013/q3/196 and his source code linked at: https://svn.nmap.org/!svn/bc/31535/nmap-exp/d33tah/lua-exec-examples/ncat/scripts/dns.lua. I stripped out everything not needed.

It basically works, i.e. it responds with our fixed ip for whatever dns query comes in. It requires nodemcu be compiled with the USE_DNS flag as mentioned above. In addition, it causes the latest 02132015 nodemcu binary to crash, (but i think that version may have restart problems with all tcp/udp servers). Its stable with the nodemcu version from 01242015 linked here https://github.com/nodemcu/nodemcu-firmware/tree/3ede2de2eae57d39f5fd810193c49254e8479fe2

Theres more work to do to wrap it with a http server and make it into full-fledged captive portal that auto-opens when connected to by an iPhone for example. 

Ive been testing it with the AP setup code project from here with some success http://www.esp8266.com/viewtopic.php?p=10852

hopefully this pushes the project forward a step or two towards a fully user-friendly AP setup mechanism.

**Comment by layerzerolabs** [comment]
Hey! that's great work mtnbrit - I'll try it out as soon as I can and get back to this thread.

**Comment by lrozema** [comment]
I have modified the DNS code to not use the unhex method, so it will work with the 0.9.5 build as well. I think it made it more efficient as well. Also an issue that only allowed to resolve once is fixed. The code is stored in my repo: https://github.com/lrozema/nodemcu-stuff/blob/master/dns.lua

**Comment by lrozema** [comment]
Perhaps useful I have put a USE_DNS build of 0.9.5 into my repo together with my webserver hack as a working captive portal. At least on my iPhone 6 it works. https://github.com/lrozema/nodemcu-stuff It seems to cause some memory leakes but it shows a proof of concept.

**Comment by cnc4less** [comment]
Irozema,

I downloaded your nodemcu-stuff but still get errors as below. any idea|

PORT OPEN 9600

Communication with MCU...
Got answer! AutoDetect firmware...

NodeMCU firmware detected.
PANIC: unprotected error in call to Lua API (webserver.lua:27: attempt to concatenate local 'path' (a nil value))
PANIC: unprotected error in call to Lua API (attempt to call a table value)
Waiting answer from ESP - Timeout reached. Command aborted.
200
path: xml3/pb3.do
302
PANIC: unprotected error in call to Lua API (webserver.lua:27: attempt to concatenate local 'path' (a nil value))
PANIC: unprotected error in call to Lua API (attempt to call a table value)
Waiting answer from ESP - Timeout reached. Command aborted.
----------------------------
dns.lc          : 1596 bytes
dns.lua         : 1484 bytes
index.html      : 338 bytes
init.lua        : 84 bytes
led.lc          : 476 bytes
led.lua         : 238 bytes
login.html      : 88 bytes
webserver.lua   : 2218 bytes
wifi_hotspot.lc : 232 bytes
wifi_hotspot.lua : 99 bytes
----------------------------
Total file(s)   : 10
Total size      : 6853 bytes

Total : 3426401 bytes
Used  : 10291 bytes
Remain: 3416110 bytes

> path: xml3/pb3.do
302
PANIC: unprotected error in call to Lua API (webserver.lua:27: attempt to concatenate local 'path' (a nil value))
PANIC: unprotected error in call to Lua API (attempt to call a table value)
ÆåÖÔô®ô@HÌ!ŠôF

**Comment by lrozema** [comment]
Are you using my build with the USE_DNS set? What client device are you trying it from?

**Comment by lighthousebulb** [comment]
Hey lrozema,

your binary looks promising, but I can't get it working.
The ESP8266 keeps rebooting after flashing the two files.
I tried it with an ESP-01 and an ESP-07.

I used the ESP8266Flasher.exe and flashed 
0x00000.bin to 0x00000
0x10000.bin to 0x10000

Do I have to check anything else or flash anything else to get it working?
Any advice?

**Comment by marcelstoer** [comment]
There are a couple of projects that attempt the same. One is this http://hackaday.com/2015/05/03/esp8266-wifi-throwies/ which uses the code from https://github.com/reischle/CaptiveIntraweb.

**Comment by lrozema** [comment]
And if you upload the nodemcu-latest from the master repo does that work?

**Comment by lighthousebulb** [comment]
Thanks so far! Finally I was able to compile the Nodemcu-firmware with USE_DNS. But I failed to upload it with the help of ESP8266Flasher.exe somehow. I finally got it uploaded with esptool.py on my Linux-Box. 
Anyway.. that works...

Now a couple thoughts to the LUA-DNS-Server
I experienced endless memory leaks. The ESP reboots after a couple of DNS-Answers with "PANIC: unprotected error in call to Lua API (not enough memory)".
You can watch "=node.heap()" getting smaller...

Looks like we have to deal with memory-leaks?
I tried the DNS-Server from https://github.com/reischle/CaptiveIntraweb and https://github.com/lrozema/nodemcu-stuff. lrozema's one seems more stable.

Looks like we are still here: https://github.com/nodemcu/nodemcu-firmware/issues/59#issuecomment-70407074

**Comment by reischle** [comment]
If using 0.9.5, my CaptiveIntraweb implementation of dns-liar.lua eats up too much heap space (using auxlib). When I  last looked, they still had "#define LUAL_BUFFERSIZE         ((BUFSIZ)*4)" in the source code which causes the problem.
The firmware I published with my code resolves that.
I have a more memory-friendly version of dns-liar.lua available and will publish that as soon as the fellow who helped me out tells me how he wants to be credited/attributed for the modification.
Still the code lacks input sanitation. You're welcome to contribute.

**Comment by TerryE** [comment]
See #719

**Comment by wiffel** [comment]
I took the code from _lrozema_ and made it even shorter. I also added _local_ to the variables and removed most of the string concatenations to save some memory.

Works fine for me. I thought I just publish it here in case somebody needs it. 

```lua

dns_ip = "\192\168\11\1";

function handleDnsRequest(con, req)
  local ix = 13;
  while req:byte(ix) > 0  do
    ix = ix + 1 + req:byte(ix);
  end
  if "\0\1" == req:sub(ix + 1, ix + 2) then
    local id    = req:sub(1, 2);
    local nr    = req:sub(5, 6);
    local query = req:sub(13, ix);
    local class = req:sub(ix + 3, ix + 4);
    local resp  = {id, "\129\128", nr, "\0\1\0\0\0\0", query,
                   "\0\1", class, "\192\12\0\1", class,
                   "\0\0\0\218\0\4", dns_ip};
    con:send(table.concat(resp));
  end
end

function setupDnsServer()
  local s = net.createServer(net.UDP);
  s:on("receive", handleDnsRequest);
  s:on("sent", function(con) con:close(); con:listen(53); end);
  s:listen(53);
end

setupDnsServer();
```

**Comment by TerryE** [comment]
I got the parsing bit down to
```Lua
local id,nr,query,class = req:match('^(..)..(..)......(' .. 
                                    ('.'):rep(ix-12) .. 
                                    ')..(..)')
if id then
  con:send(('%s\129\128%s\0\1\0\0\0\0%s\0\1%s\192\12\0\1%s\0\0\0\218\0\4'):
      format(id,nr,query,class,class))
end
```
(untested) :smile:

**Comment by joysfera** [comment]
@TerryE , FYI, your super short version does not work correctly, unfortunately. Seems like the binary zeros cause some issues both in req:match() and string:format() strings.

**Comment by TerryE** [comment]
@joysfera Petr you might be correct the binary zeros on these calls.  However, in this case @pjsg has now added the struct module which could be used as an alternative to `string,sub()` picking and Lua optimises chained string concatenation, something like:
```Lua
local id,nr,query,check,class = struct.unpack("c2x2c2x5c"..(ix-12).."i2c2", rec)
```
and send if `check` is 1.  

In terms of runtime cost  `table.concat` variant generates more code (19 instructions, including a table creation and the two calls).  A straight chained concatenation is faster and generates less code (19 instructions)  -- Compare the two version using luac or luac.cross.

By chained concat I mean something like `con:send(a..b..c..d..e)` where the elements can be variables or constants, and not the style
```Lua
resp = a
resp = resp..b
resp = resp..c
resp = resp..d
resp = resp..e
cond:send(resp)
```
which is horribly inefficient in terms of both code size and runtime in Lua

**Comment by joysfera** [comment]
Great suggestions, thank you @TerryE ! Pity I don't have the struct module compiled in my existing firmware, seems to be a very useful addition.
I'll give the chained concat a try, thanks.

**Comment by whendonkiesfly** [comment]
I can confirm the issues with \0 %s replacement in string.format. Is this considered a nodemcu bug? I have been affected by the same issue with the print function.

**Comment by TerryE** [comment]
> I can confirm the issues with \\0 %s replacement in string.format. Is this considered a nodemcu bug? I have been affected by the same issue with the print function.

My point is that if you type **print "hello\0world"** on the PC in bog standard Lua 5.1 then you will see **hello** printed out. 5.3 doesn't do this, but 5.1 does. We will fix this with the 5.3 upgrade but not with 5.1 because this breaks backward compatibility.
